src/sync/sync_session.cpp

```diff
@@ -379,23 +360,17 @@ void SyncSession::handle_error(SyncError error)
                 // when the old session is torn down, which we don't want as this
                 // is supposed to be transparent to the user.
                 //
-                // To avoid this, we need to do two things: move the completion
-                // handlers aside temporarily so that moving to the inactive
-                // state doesn't clear them, and track which sync::Session each
-                // completion notification came from so that we can ignore
-                // notifications from the old session.
+                // To avoid this, we need to move the completion handlers aside temporarily so
+                // that moving to the inactive state doesn't clear them - they will be re-registered
+                // when the session becomes active again.
                 {
                     std::unique_lock<std::mutex> lock(m_state_mutex);
                     m_force_client_resync = true;
 
-                    ++m_client_resync_counter;
-                    auto download_handlers = std::move(m_download_completion_callbacks);
-                    auto upload_handlers = std::move(m_upload_completion_callbacks);
-
+                    auto completion_callbacks = std::move(m_completion_callbacks);
```

Comment:
In the other spots we do this you changed it to swapping with a local which is probably better, so this one should as well.

Reply:
Will do. I also added a FIXME comment since I realized advance_state is not exception safe so we should make sure that we always move the callbacks back in place by using a guard type, but I don't think we need to fix that in this PR.